core/src/main/java/com/rudderstack/android/sdk/core/RudderDataResidency.java

```diff
@@ -0,0 +1,130 @@
+package com.rudderstack.android.sdk.core;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import java.util.Map;
+
+class RudderDataResidency {
+    private final RudderConfig config;
+    @VisibleForTesting
+    Map<String, String> dataResidencyUrls = null;
+    @VisibleForTesting
+    ResidencyServer residencyServer;
+
+    RudderDataResidency(@Nullable RudderServerConfig serverConfig, @NonNull RudderConfig config) {
+        this.config = config;
+        if (serverConfig != null && serverConfig.source != null && serverConfig.source.dataResidencyUrls != null) {
+            this.dataResidencyUrls = serverConfig.source.dataResidencyUrls;
+        }
+        this.residencyServer = config.getResidencyServer();
+    }
+
+    /**
+     * This API will update the dataPlane URL present in the config, just after sourceConfig was fetched successfully.
+     *
+     * <p>Preference of dataPlane URL is decided as follows:</p>
+     * <ul>
+     *     <li>First look for the Residency server (which was set while SDK initialisation) in the source config.</li>
+     *     <li>If Residency server was not set while SDK initialisation then fallback to default US residency server.</li>
+     *     <li>If residency server list is empty or null then fallback to dataPlane URL (which was set while SDK initialisation).</li>
+     *     <li>If dataPlane URL was not passed while SDK initialisation then as a last resort fallback to the default dataPlane URL.</li>
+     * </ul>
+     * <p>
+     * Default Residency server is US.
+     * <p>
+     * Use the below code snippet to set the Data Residency:
+     * <pre>
+     *     {@code
+     *     rudderClient = RudderClient.getInstance(
+     *             this,
+     *             WRITE_KEY,
+     *             RudderConfig.Builder()
+     *                 .withResidencyServer(ResidencyServer.US)
+     *                 .build()
+     *     )
+     *      }
+     * </pre>
+     */
+    void handleDataPlaneUrl() {
+        if (residencyServer == ResidencyServer.US) {
+            handleDefaultServer();
+        } else {
+            handleOtherServer(residencyServer);
+        }
+    }
+
+    /**
+     * Handle all server other than US.
+     * If passed server is not present then fallback to US server.
+     *
+     * @param residencyServer Residency server set while SDK initialisation
+     */
+    @VisibleForTesting
+    void handleOtherServer(@NonNull ResidencyServer residencyServer) {
+        // TODO: Decide upper or lower case
+        String dataResidencyUrl = getDataResidencyUrl(residencyServer.name());
+        if (dataResidencyUrl != null) {
+            config.setDataPlaneUrl(dataResidencyUrl);
```

Comment:
I would suggest not updating the config object with the final result data plane URL and letting it be whatever the user has passed.  
Let's calculate it based on the ServerConfig downloaded, and RudderConfig supplied by the user and use it wherever needed.

Reply:
Implemented the change in this PR: https://github.com/rudderlabs/rudder-sdk-android/pull/150.
Please review it.